test/one/profiler/test/TestProcess.java

```diff
@@ -0,0 +1,302 @@
+import java.util.concurrent.TimeoutException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Stream;
+
+import one.jfr.JfrReader;
+
+public class TestProcess implements Closeable {
+    private static final Logger log = Logger.getLogger(TestProcess.class.getName());
+
+    private static final Pattern filePattern = Pattern.compile("(%[a-z]+)(\\.[a-z]+)?");
+
+    private static final MethodHandle pid = getPidHandle();
+
+    private static MethodHandle getPidHandle() {
+        // JDK 9+
+        try {
+            return MethodHandles.publicLookup().findVirtual(Process.class, "pid", MethodType.methodType(long.class));
+        } catch (ReflectiveOperationException e) {
+            // fallback
+        }
+
+        // JDK 8
+        try {
+            Field f = Class.forName("java.lang.UNIXProcess").getDeclaredField("pid");
+            f.setAccessible(true);
+            return MethodHandles.lookup().unreflectGetter(f).asType(MethodType.methodType(long.class, Process.class));
+        } catch (ReflectiveOperationException e) {
+            throw new IllegalStateException("Unsupported API", e);
+        }
+    }
+
+    private final Process p;
+    private final Map<String, File> tmpFiles = new HashMap<>();
+    private final int timeout = 30;
+    private String testName = null;
+
+    public TestProcess(Test test, String libExt, String testName) throws Exception {
+        this(test, libExt);
+        this.testName = testName; // if testName is not null then we retain the logs into that folder
+    }
+
+    public TestProcess(Test test, String libExt) throws Exception {
+        List<String> cmd = new ArrayList<>();
+        cmd.add(System.getProperty("java.home") + "/bin/java");
+        cmd.add("-cp");
+        cmd.add(System.getProperty("java.class.path"));
+        if (test.debugNonSafepoints()) {
+            cmd.add("-XX:+UnlockDiagnosticVMOptions");
+            cmd.add("-XX:+DebugNonSafepoints");
+        }
+        addArgs(cmd, test.jvmArgs());
+        if (!test.agentArgs().isEmpty()) {
+            cmd.add("-agentpath:build/lib/libasyncProfiler." + libExt + "=" + substituteFiles(test.agentArgs()));
+        }
+        cmd.add(test.mainClass().getName());
+        addArgs(cmd, test.args());
+        log.log(Level.FINE, "Running " + cmd);
+        ProcessBuilder pb = new ProcessBuilder(cmd).inheritIO();
+        if (test.output()) {
+            pb.redirectOutput(createTempFile("%out"));
+        }
+        if (test.error()) {
+            pb.redirectError(createTempFile("%err"));
+        }
+        if (!test.agentArgs().isEmpty() || test.jvmArgs().contains("-Djava.library.path=build/lib")) {
+            pb.redirectOutput(createTempFile("%pout"));
+            pb.redirectError(createTempFile("%perr"));
+        }
+
+        this.p = pb.start();
+        // Give the JVM some time to initialize
+        Thread.sleep(700);
```

Comment:
Better to replace with something more determenistic. We may discuss what exactly.